datastore/src/json_file.rs

```diff
@@ -0,0 +1,235 @@
+            IoErrorKind::Other,
+            "couldn't get parent directory of destination",
+        ))?;
+        let mut temporary_file = NamedTempFile::new_in(self_path_parent)?;
+
+        let content = self.content.lock();
+        to_writer(&mut temporary_file, &*content)?;
+        temporary_file.sync_data()?;
+
+        // Note that `persist` will fail if we try to persist across filesystems. However that
+        // shouldn't happen since we created the temporary file in the same directory as the final
+        // path.
+        temporary_file.persist(&self.path)?;
+        Ok(())
+    }
+}
+
+impl Datastore for JsonFileDatastore {
+    fn put(&self, key: Cow<str>, value: Vec<u8>) {
+        let mut content = self.content.lock();
+        content.insert(key.into_owned(), Value::String(base64::encode(&value)));
+    }
+
+    fn get(&self, key: &str) -> Option<Vec<u8>> {
+        let content = self.content.lock();
+        // If the JSON is malformed, we just ignore the value.
+        content.get(key).and_then(|val| match val {
+            &Value::String(ref s) => base64::decode(s).ok(),
+            _ => None,
+        })
+    }
+
+    fn has(&self, key: &str) -> bool {
+        let content = self.content.lock();
+        content.contains_key(key)
+    }
+
+    fn delete(&self, key: &str) -> bool {
+        let mut content = self.content.lock();
+        content.remove(key).is_some()
+    }
+
+    fn query<'a>(
+        &'a self,
+        query: Query,
+    ) -> Box<Stream<Item = (String, Vec<u8>), Error = IoError> + 'a> {
+        let content = self.content.lock();
+
+        let keys_only = query.keys_only;
+
+        let content_stream = iter_ok(content.iter().filter_map(|(key, value)| {
+            // Skip values that are malformed.
+            let value = if keys_only {
+                Vec::with_capacity(0)
+            } else {
+                match value {
+                    &Value::String(ref s) => {
+                        match base64::decode(s) {
+                            Ok(s) => s,
+                            Err(_) => return None,
+                        }
+                    }
+                    _ => return None,
+                }
+            };
+
+            Some((key.clone(), value))
+        }));
+
+        // `content_stream` reads from the content of the `Mutex`, so we need to clone the data
+        // into a `Vec` before returning.
+        let collected = naive_apply_query(content_stream, query)
+            .collect()
+            .wait()
+            .expect("can only fail if either `naive_apply_query` or `content_stream` produce \
+                     an error, which cann't happen");
+        let output_stream = iter_ok(collected.into_iter());
+        Box::new(output_stream) as Box<_>
+    }
+}
+
+impl Drop for JsonFileDatastore {
+    #[inline]
+    fn drop(&mut self) {
+        let _ = self.flush(); // What do we do in case of error? :-/
```

Comment:
Why would it error? Like disk-access problems? If there's nothing we can do, at least put in some documentation under which conditions it could fail and what would happen to data upon failure.